object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java

```diff
@@ -123,129 +119,89 @@ void checkMustCallInvoked(ControlFlowGraph cfg) {
-    // to correctly handle foreach loops.
-    boolean assignedToOwning = !node.getInSource();
-    AssignmentContext assignmentContext = node.getAssignmentContext();
-    if (assignmentContext != null) {
-      Element elementForType = assignmentContext.getElementForType();
-      if (assignmentContext instanceof AssignmentLhsContext) {
-        // lhs should be a local variable
-        assignedToOwning = isOwningAssignmentLhs(elementForType);
-      } else if (assignmentContext instanceof MethodParameterContext) {
-        // must be an @Owning or @MustCallChoice parameter
-        assignedToOwning =
-            typeFactory.getDeclAnnotation(elementForType, Owning.class) != null
-                || typeFactory.hasMustCallChoice(elementForType);
-      } else if (assignmentContext instanceof MethodReturnContext) {
-        // must be an @Owning return
-        assignedToOwning =
-            TRANSFER_OWNERSHIP_AT_RETURN
-                ? typeFactory.getDeclAnnotation(
-                        TreeUtils.elementFromTree(assignmentContext.getContextTree()),
-                        NotOwning.class)
-                    == null
-                : typeFactory.getDeclAnnotation(elementForType, Owning.class) != null;
-      } else if (assignmentContext instanceof LambdaReturnContext) {
-        // TODO handle this case.  For now we will report an error
-      } else {
-        throw new BugInCF("unexpected AssignmentContext type " + assignmentContext.getClass());
+    if (typeFactory.mapTempVarToNode.inverse().containsKey(node.getTree())) {
+
+      LocalVariableNode temporaryLocal = typeFactory.mapTempVarToNode.inverse().get(node.getTree());
+      LocalVarWithTree lhsLocalVarWithTreeNew =
+          new LocalVarWithTree(new LocalVariable(temporaryLocal), tree);
+
+      Node receiver = null;
+      if (node instanceof ObjectCreationNode || node instanceof MethodInvocationNode) {
+        receiver = getVarOrTempVarPassedAsMustCallChoiceParam(node);
       }
-    }
-    if (!assignedToOwning) {
-      // check if @CalledMethods type of return satisfies the @MustCall obligation
-      AnnotationMirror cmAnno =
-          typeFactory.getAnnotatedType(tree).getAnnotationInHierarchy(typeFactory.top);
-      if (!calledMethodsSatisfyMustCall(mustCallVal, cmAnno)) {
-        checker.reportError(
-            tree,
-            "required.method.not.called",
-            MustCallInvokedChecker.formatMissingMustCallMethods(mustCallVal),
-            TreeUtils.typeOf(tree).toString(),
-            "never assigned to an @Owning location");
+
+      if (receiver == null
+          && node instanceof MethodInvocationNode
+          && (typeFactory.returnsThis((MethodInvocationTree) tree))) {
+        receiver = ((MethodInvocationNode) node).getTarget().getReceiver();
+        if (receiver instanceof MethodInvocationNode) {
+          receiver = typeFactory.mapTempVarToNode.inverse().get(receiver.getTree());
```

Comment:
We should add a method `getTempVarForTree()` to the type factory that encapsulates this logic of doing a `get()` on the `inverse()`, and use that everywhere